### PR TITLE
refactor pyproto internals

### PR DESCRIPTION
**Pull request by dvermd**
This PR aims to refactor MethodProto.

`make test` passed OK

Still to do:
- [ ] an entry in CHANGELOG.md
- [x] `Free` methods 

Closes: #1117

### CODE REVIEW

File: pyo3-derive-backend/src/utils.rs [hunk 1]
```diff
@@ -1,12 +1,7 @@
 // Copyright (c) 2017-present PyO3 Project and Contributors
 use proc_macro2::Span;
-use proc_macro2::TokenStream;
 use std::fmt::Display;
 
-pub fn print_err(msg: String, t: TokenStream) {
```
Comment (kngwyu) [inline pyo3-derive-backend/src/utils.rs:6]: Good catch :+1:

**Comment by kngwyu** [comment]
Quite a nice refactoring, thanks.
I think your approach to `Free` methods is correct. They doesn't have `type Result = ...`.

How about adding some constructors to `MethodProto`? For example, `with_result: false` is only used for `GC` methods, so we can use a 'short-cut' constructor for the common case.

**Comment by kngwyu** [comment]
BTW, I came up with a shorter `modify_arg_ty`:
```rust

fn modify_arg_ty(
    sig: &mut syn::Signature,
    idx: usize,
    decl1: &syn::FnArg,
    decl2: &syn::FnArg,
) -> syn::Result<()> {
    let arg = sig.inputs[idx].clone();
    match arg {
        syn::FnArg::Typed(ref cap) if crate::utils::option_type_argument(&*cap.ty).is_some() => {
            sig.inputs[idx] = fix_name(&cap.pat, &decl2.inputs[idx])?;
        }
        syn::FnArg::Typed(ref cap) => {
            sig.inputs[idx] = fix_name(&cap.pat, &decl1.inputs[idx])?;
        }
        _ => return Err(syn::Error::new_spanned(arg, "not supported")),
    }

    Ok(())
}
```

**Comment by dvermd** [comment]
> How about adding some constructors to `MethodProto`? For example, `with_result: false` is only used for `GC` methods, so we can use a 'short-cut' constructor for the common case.

I thought and tried some constructors but then the declarations in `defs.rs` won't be that explicit without constructors' arguments names. If you're OK with that I can create them.

Else, there's the builder pattern that can be a solution `builder(name, proto).args(&[...]).with_self().with_result()`.

Which one do you like better ?

File: pyo3-derive-backend/src/proto_method.rs [hunk 2]
```diff
@@ -1,70 +1,59 @@
...
+            args: &[],
+            with_self: false,
+            with_result: true,
+        }
+    }
+    pub const fn args(self, args: &'static [&'static str]) -> MethodProtoBuilder {
+        let mut with_args = self;
+        with_args.args = args;
+        with_args
+    }
```
Comment (davidhewitt) [inline pyo3-derive-backend/src/proto_method.rs:39]: I think you can probably take `mut self` in each of these builder methods, e.g.

```suggestion
    pub const fn args(mut self, args: &'static [&'static str]) -> MethodProtoBuilder {
        self.args = args;
        self
    }
```

File: pyo3-derive-backend/src/pyproto.rs [hunk 3]
```diff
@@ -66,6 +66,7 @@ fn impl_proto_impl(
     for iimpl in impls.iter_mut() {
         if let syn::ImplItem::Method(ref mut met) = iimpl {
             // impl Py~Protocol<'p> { type = ... }
+            // TODO: MethodProto.result = meth.sig.output
```
Comment (davidhewitt) [inline pyo3-derive-backend/src/pyproto.rs:69]: Not sure what this TODO is asking to happen?

**Review by davidhewitt — APPROVED** [review]
🚀 Thanks, this looks awesome to me! Lovely to see a great reduction in LOC and overall it's much easier to understand the affected files.

The `.has_self()` builder method is really nice as it'll make it super easy to refactor the self-types later as discussed on the issue.

Regarding CHANGELOG entry, I don't think this affects users so imo it shouldn't have an entry.

**Comment by davidhewitt** [comment]
Regarding MSRV failure - I think there's now maybe enough reasons for us to bump to rust 1.45, so I'm probably going to open a PR for that later.

File: pyo3-derive-backend/src/pyproto.rs (hunk 3 above)
Comment (dvermd) [reply pyo3-derive-backend/src/pyproto.rs]: This TODO was forgotten there. It's removed.

File: pyo3-derive-backend/src/proto_method.rs (hunk 2 above)
Comment (dvermd) [reply pyo3-derive-backend/src/proto_method.rs]: You're right, I didn't think `mut self` was possible but the code is much nicer this way.

**Comment by davidhewitt** [comment]
Looks great, thanks! If you squash and rebase on master I think you should find that MSRV CI starts working.

I notice this PR is still in draft, but I think other than a rebase it's now hopefully ready to be merged?

**Comment by dvermd** [comment]
Squashed, rebased, removed a `use` comment in `defs.rs` and un-drafted.

Still some problems with MSRV but 1.45 should contain the right commit

**Comment by davidhewitt** [comment]
Ugh. 

Based on [this comment](https://github.com/rust-lang/rust/pull/73862/#issue-441477879)

>  This PR stabilizes these casts (which are already stable in const outside const fn)

I wonder if this works as a workaround:

```rust

// TODO: workaround for no unsized casts in const fn on Rust 1.45 (stable in 1.46)
const EMPTY_ARGS: &[&str] = &[];

impl MethodProtoBuilder {
        MethodProtoBuilder {
            name,
            proto,
            args: EMPTY_ARGS,
            with_self: false,
            with_result: true,
        }
    }
}
```

If you don't have a chance before I do, I might push a commit to this branch tonight to try that.

**Comment by davidhewitt** [comment]
... or I guess if that doesn't work, maybe could just always take `args` as an input to the `const fn new`.

**Comment by kngwyu** [comment]
Just a nit, but your `EMPTY_ARGS` can be an associated constant, like:
```rust
impl MethodProtoBuilder {
    const EMPTY_ARGS: &[&str] = &[];
...
}
```

**Comment by kngwyu** [comment]
Why use `build`? 
It looks like that `MethodProtoBuilder` and `MethodProto` is the same and I don't see any reason for defining the builder as another struct.

**Comment by dvermd** [comment]
> ```rust
> impl MethodProtoBuilder {
>     const EMPTY_ARGS: &[&str] = &[];
> ...
> }
> ```
This workaround works !

> Why use `build`?
> It looks like that `MethodProtoBuilder` and `MethodProto` are the same and I don't see any reason for defining the builder as another struct.

I merged the `MethodProtoBuilder` into the `MethodProto` and removed the `build()` method.

**Comment by davidhewitt** [comment]
> It looks like that MethodProtoBuilder and MethodProto are the same and I don't see any reason for defining the builder as another struct.

👍 good spot

> I merged the MethodProtoBuilder into the MethodProto and removed the build() method.

🎉 this refactoring is really great; it's reduced LOC by just over 50% 2 for this part of the codebase. Thanks so much for your work on this!

**Pull request merged**